sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala

```diff
@@ -74,9 +75,18 @@ private[sql] class HiveSessionCatalog(
   // | Methods and fields for interacting with HiveMetastoreCatalog |
   // ----------------------------------------------------------------
 
-  override def getDefaultDBPath(db: String): String = {
-    val defaultPath = context.hiveconf.getVar(HiveConf.ConfVars.METASTOREWAREHOUSE)
-    new Path(new Path(defaultPath), db + ".db").toString
+  // This function is to get the path for creating a non-default database
+  override def getDatabasePath(dbName: String, path: Option[String] = None): String = {
+    val dbPath = path.map(new Path(_)).getOrElse {
+      val defaultPath = context.hiveconf.getVar(HiveConf.ConfVars.METASTOREWAREHOUSE)
+      if (StringUtils.isBlank(defaultPath)) {
+        throw new AnalysisException(
+          s"${HiveConf.ConfVars.METASTOREWAREHOUSE.varname} is not set in the config or blank")
+      }
+      new Path(new Path(defaultPath), dbName.toLowerCase() + ".db")
+    }
+    val fs = dbPath.getFileSystem(context.hiveconf)
+    new Path(fs.getUri.getScheme, fs.getUri.getAuthority, dbPath.toUri.getPath).toString
```

Comment:
Sure, will add comments and rename it. 

Here, I am just following what metastore is doing for path generation. I think the major reason is documented: https://github.com/apache/hive/blob/master/metastore/src/java/org/apache/hadoop/hive/metastore/Warehouse.java#L127-L133

The paths in `HiveConf.ConfVars.METASTOREWAREHOUSE` or user-provided location could have potential issues. I think that might be the reason why metastore did it like https://github.com/apache/hive/blob/master/metastore/src/java/org/apache/hadoop/hive/metastore/Warehouse.java#L159